homeassistant/components/light/tradfri.py

```diff
@@ -70,49 +93,68 @@ def brightness(self):
+        keys = {}
+        if ATTR_TRANSITION in kwargs:
+            keys['transition_time'] = int(kwargs[ATTR_TRANSITION])
+
         if ATTR_BRIGHTNESS in kwargs:
-            self._api(self._group.set_dimmer(kwargs[ATTR_BRIGHTNESS]))
+            self.hass.async_add_job(self._api(
+                self._group.set_dimmer(kwargs[ATTR_BRIGHTNESS], **keys)))
         else:
-            self._api(self._group.set_state(1))
+            self.hass.async_add_job(self._api(self._group.set_state(1)))
+
+    @callback
+    def _async_start_observe(self, exc=None):
+        """Start observation of light."""
+        from pytradfri.error import PyTradFriError
+        if exc:
+            _LOGGER.warning("Observation failed for %s", self._name,
+                            exc_info=exc)
 
-    def update(self):
-        """Fetch new state data for this group."""
-        from pytradfri import RequestTimeout
         try:
-            self._api(self._group.update())
-        except RequestTimeout:
-            _LOGGER.warning("Tradfri update request timed out")
+            cmd = self._group.observe(callback=self._observe_update,
+                                      err_callback=self._async_start_observe,
+                                      duration=0)
+            self.hass.async_add_job(self._api(cmd))
+        except PyTradFriError as err:
+            _LOGGER.warning("Observation failed, trying again", exc_info=err)
+            self._async_start_observe()
+
+    def _refresh(self, group):
+        """Refresh the light data."""
+        self._group = group
+        self._name = group.name
+
+    def _observe_update(self, tradfri_device):
+        """Receive new state data for this light."""
+        self._refresh(tradfri_device)
+
+        self.hass.async_add_job(self.async_update_ha_state())
 
 
-class Tradfri(Light):
-    """The platform class required by Home Asisstant."""
+class TradfriLight(Light):
+    """The platform class required by Home Assistant."""
 
     def __init__(self, light, api):
         """Initialize a Light."""
-        self._light = api(light)
         self._api = api
-
-        # Caching of LightControl and light object
-        self._light_control = self._light.light_control
-        self._light_data = self._light_control.lights[0]
-        self._name = self._light.name
+        self._light = None
+        self._light_control = None
+        self._light_data = None
+        self._name = None
         self._rgb_color = None
-        self._features = SUPPORT_BRIGHTNESS
+        self._features = SUPPORTED_FEATURES
+        self._temp_supported = False
 
-        if self._light_data.hex_color is not None:
-            if self._light.device_info.manufacturer == IKEA:
-                self._features |= SUPPORT_COLOR_TEMP
-            else:
-                self._features |= SUPPORT_RGB_COLOR
-
-        self._ok_temps = \
-            self._light.device_info.manufacturer in ALLOWED_TEMPERATURES
+        self._refresh(light)
```

Comment:
I think that should be call inside `async_add_to_hass`. A init can be run on thread or loop

Reply:
It's not actually doing a refresh here, just setting instance variables.